cli/internal/oauth/api/api_test.go

```diff
@@ -0,0 +1,238 @@
+			r.ParseForm()
+			clientID = r.FormValue("client_id")
+			audience = r.FormValue("audience")
+			scope = r.FormValue("scope")
+			path = r.URL.Path
+
+			jsonState, err := json.Marshal(expectedState)
+			assert.NilError(t, err)
+
+			_, _ = w.Write(jsonState)
+		}))
+		defer ts.Close()
+		api := API{
+			BaseURL:  ts.URL,
+			ClientID: "aClientID",
+			Scopes:   []string{"bork", "meow"},
+			Client:   util.Client{},
+		}
+
+		state, err := api.GetDeviceCode(context.Background(), "anAudience")
+		assert.NilError(t, err)
+
+		assert.DeepEqual(t, expectedState, state)
+		assert.Equal(t, clientID, "aClientID")
+		assert.Equal(t, audience, "anAudience")
+		assert.Equal(t, scope, "bork meow")
+		assert.Equal(t, path, "/oauth/device/code")
+	})
+
+	t.Run("error w/ description", func(t *testing.T) {
+		ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			jsonState, err := json.Marshal(TokenResponse{
+				ErrorDescription: "invalid audience",
+			})
+			assert.NilError(t, err)
+
+			w.WriteHeader(http.StatusBadRequest)
+			_, _ = w.Write(jsonState)
+		}))
+		defer ts.Close()
+		api := API{
+			BaseURL:  ts.URL,
+			ClientID: "aClientID",
+			Scopes:   []string{"bork", "meow"},
+			Client:   util.Client{},
+		}
+
+		_, err := api.GetDeviceCode(context.Background(), "bad_audience")
+
+		assert.ErrorContains(t, err, "invalid audience")
+	})
+
+	t.Run("general error", func(t *testing.T) {
+		ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			http.Error(w, "an error", http.StatusInternalServerError)
+		}))
+		defer ts.Close()
+		api := API{
+			BaseURL:  ts.URL,
+			ClientID: "aClientID",
+			Scopes:   []string{"bork", "meow"},
+			Client:   util.Client{},
+		}
+
+		_, err := api.GetDeviceCode(context.Background(), "anAudience")
+
+		assert.ErrorContains(t, err, "failed to get device code")
+	})
+
+	t.Run("canceled context", func(t *testing.T) {
+		ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			time.Sleep(2 * time.Second)
+			http.Error(w, "an error", http.StatusInternalServerError)
+		}))
+		defer ts.Close()
+		api := API{
+			BaseURL:  ts.URL,
+			ClientID: "aClientID",
+			Scopes:   []string{"bork", "meow"},
+			Client:   util.Client{},
+		}
+
+		ctx, cancel := context.WithCancel(context.Background())
+		go func() {
+			time.Sleep(1 * time.Second)
+			cancel()
+		}()
+		_, err := api.GetDeviceCode(ctx, "anAudience")
+
+		assert.ErrorContains(t, err, "context canceled")
+	})
+}
+
+func TestWaitForDeviceToken(t *testing.T) {
+	t.Run("success", func(t *testing.T) {
+		expectedToken := TokenResponse{
+			AccessToken:  "a-real-token",
+			IDToken:      "",
+			RefreshToken: "the-refresh-token",
+			Scope:        "",
+			ExpiresIn:    3600,
+			TokenType:    "",
+		}
+		var respond bool
+		go func() {
+			time.Sleep(10 * time.Second)
+			respond = true
+		}()
```

Comment:
Yeah! I thought it'd be good to keep a test with "somewhat realistic" settings, and it might take a user a few seconds to log in. Will mark it parallel though.